crates/ruff_python_parser/src/parenthesize.rs

```diff
@@ -0,0 +1,270 @@
+use ruff_python_ast::{AnyNodeRef, ExprRef};
+use ruff_text_size::{Ranged, TextLen, TextRange};
+
+use crate::{TokenKind, Tokens};
+
+/// Tokens that should be treated as trivia when scanning around parentheses.
+/// Mirrors the behavior of `SimpleTokenKind::is_trivia()` as closely as possible
+/// at the `TokenKind` level.
+const fn is_trivia(kind: TokenKind) -> bool {
+    matches!(
+        kind,
+        TokenKind::Comment
+            | TokenKind::Newline
+            | TokenKind::NonLogicalNewline
+            | TokenKind::Indent
+            | TokenKind::Dedent
+    )
+}
+
+/// Returns an iterator over the ranges of the optional parentheses surrounding an expression.
+///
+/// E.g. for `((f()))` with `f()` as expression, the iterator returns the ranges (1, 6) and (0, 7).
+///
+/// Note that without a parent the range can be inaccurate, e.g. `f(a)` we falsely return a set of
+/// parentheses around `a` even if the parentheses actually belong to `f`. That is why you should
+/// generally prefer [`parenthesized_range`].
+pub fn parentheses_iterator<'a>(
+    expr: ExprRef<'a>,
+    parent: Option<AnyNodeRef>,
+    tokens: &'a Tokens,
+) -> impl Iterator<Item = TextRange> + 'a {
+    let exclusive_parent_end = if let Some(parent) = parent {
+        // If the parent is a node that brings its own parentheses, exclude the closing parenthesis
+        // from our search range. Otherwise, we risk matching on calls, like `func(x)`, for which
+        // the open and close parentheses are part of the `Arguments` node.
+        if parent.is_arguments() {
+            parent.end() - ")".text_len()
+        } else {
+            parent.end()
+        }
+    } else {
+        tokens.last().map_or(expr.end(), Ranged::end)
+    };
+
+    let right_parens = tokens
+        .after(expr.end())
+        .iter()
+        .take_while(move |token| token.start() < exclusive_parent_end)
+        .filter(|token| !is_trivia(token.kind()))
+        .take_while(|token| token.kind() == TokenKind::Rpar);
```

Comment:
Nit: You can rewrite it like this to keep it closer to the original implementation:

```suggestion
    let after_tokens = if let Some(parent) = parent {
        // If the parent is a node that brings its own parentheses, exclude the closing parenthesis
        // from our search range. Otherwise, we risk matching on calls, like `func(x)`, for which
        // the open and close parentheses are part of the `Arguments` node.
        let exclusive_parent_end = if parent.is_arguments() {
            parent.end() - ")".text_len()
        } else {
            parent.end()
        };

        tokens.in_range(TextRange::new(expr.end(), exclusive_parent_end))
    } else {
        tokens.after(expr.end())
    };
    
    let right_parens = after_tokens
        .iter()
        .filter(|token| !token.kind().is_trivia())
        .take_while(move |token| token.kind() == TokenKind::Rpar);
```